### DISCUSSION TITLE
App breaks after migrating to v13

### DISCUSSION THREAD

**Discussion by Eliepse** [vercel/next.js#44730] [Help] [answered]
Hello,
I have my codebase running perfectly on Nextjs 12, but if I upgrade to the version 13 I can't print any page. The node process just breaks. I'm on nodejs 18.13.0.

I didn't switch to the experimental "appDir" and stayed on the "pages" folder. I have Redux set up on the _app.js, where I also included some logic to run on the front-end.

I tried to add "use client" on top of the files of components used it this page, but doesn't seem to help much. The project I'm working on is already quite big so I can't start fresh and re-import everything. It looks like it's a client/server component error, yet it should be compatible according Nextjs doc as I didn't moved to the experimental "appDir".

Here is the output in the console

```
$ next dev -p 3002
ready - started server on 0.0.0.0:3002, url: http://localhost:3002
info  - Loaded env from C:\Users\micro\www\admin\.env.development
info  - Loaded env from C:\Users\micro\www\admin\.env
event - compiled client and server successfully in 10.2s (3534 modules)
wait  - compiling...
event - compiled successfully in 2.1s (3325 modules)
wait  - compiling /test (client and server)...
event - compiled client and server successfully in 521 ms (3537 modules)
node:internal/child_process:821
        handle.setSimultaneousAccepts(true);
               ^

TypeError: handle.setSimultaneousAccepts is not a function
    at target._send (node:internal/child_process:821:16)
    at target.send (node:internal/child_process:738:19)
    at sendHelper (node:internal/cluster/utils:28:15)
    at send (node:internal/cluster/primary:348:10)
    at node:internal/cluster/primary:314:5
    at SharedHandle.add (node:internal/cluster/shared_handle:30:3)
    at queryServer (node:internal/cluster/primary:308:10)
    at Worker.onmessage (node:internal/cluster/primary:254:5)
    at ChildProcess.onInternalMessage (node:internal/cluster/utils:49:5)
    at ChildProcess.emit (node:events:525:35)

Node.js v18.13.0
error Command failed with exit code 1.
```

Where am I wrong? I could understand having an error on the browser, not directly on the console like that...

**Comment by icyJoseph** [comment]
Did you also upgrade your Node version? Have you tried switching Node.js versions? This is really strange.  Are you using pm2 or something like that?

**Reply by icyJoseph** [reply]
I found this issue - https://github.com/mikaelbr/node-notifier/issues/410#issue-1312986853 - which has the exact same error printed out, and it is also on windows. I wonder if Next 13 has upgraded `node-notifier` in some way, or through a dependency perhaps ~ gotta think a bit about this one.

**Reply by Eliepse** [reply]
Thank you for your answer. I was on node 16 before and updated because of the error. And yes I'm on Windows as I forgot to mention it.

Interesting lead with node-notifier! I wasn't able to find it.

**Reply by icyJoseph** [reply]
Yeah, but I am not sure it is used by next.js anymore. Perhaps you have it in your project? You can do `npm explain node-notifier` and see if its there. I checked the compiled  dependencies and it wasn't there, and in fact, it is nowhere to be found in the repo anymore...

**Reply by icyJoseph** [reply]
https://github.com/vercel/next.js/pull/43801 Yeah it was removed on december. What version of Next are you using? and please also run `npm explain node-notifier` perhaps you have it as part of another dependency.

**Reply by Eliepse** [reply]
Hello, sorry for the delay I wasn't working yesterday.

So, I checked node-notifier and it seems that it's not to be found on the project. My project fails on Nextjs 13.0 and 13.1.1.
I continue the investigation to find out which package causes this issue.

**Comment by Eliepse** [comment]
Ok, I found a lead!
I found a way to have a bit more stack trace and the problem is related to memory analysis of Next.

```
target._send             	child_process:821
target.send              	child_process:738
sendHelper               	utils:28
send                     	primary:348
(anonymous function)     	primary:314
SharedHandle.add         	shared_handle:30
queryServer              	primary:308
onmessage                	primary:254
onInternalMessage        	utils:49
emit                     	events:525
emit                     	child_process:937
processTicksAndRejections	task_queues:83
Async call from TickObject
init                     	inspector_async_hook:25
emitInitNative           	async_hooks:202
emitInitScript           	async_hooks:505
nextTick                 	task_queues:132
handleMessage            	child_process:953
channel.onread           	child_process:619
callbackTrampoline       	async_hooks:130
Async call from PIPEWRAP
init                     	inspector_async_hook:25
emitInitNative           	async_hooks:202
(anonymous function)     	child_process:1028
getValidStdio            	child_process:991
ChildProcess.spawn       	child_process:355
spawn                    	child_process:757
fork                     	child_process:170
createWorkerProcess      	primary:129
cluster.fork             	primary:164
nextDev                  	next-dev.js:321
(anonymous function)     	next.ts:122
Async call from Promise.then
(anonymous function)     	next.ts:122
Module._compile          	loader:1218
Module._extensions..js   	loader:1272
Module.load              	loader:1081
Module._load             	loader:922
executeUserEntryPoint    	run_main:81
(anonymous function)     	run_main_module:23
```

```js
// next-dev.js
// ...

// we're using a sub worker to avoid memory leaks. When memory usage exceeds 90%, we kill the worker and restart it.
// this is a temporary solution until we can fix the memory leaks.
// the logic for the worker killing itself is in `packages/next/server/lib/start-server.ts`
if (!process.env.__NEXT_DISABLE_MEMORY_WATCHER && _cluster.default.isMaster) {
     _cluster.default.fork();
    _cluster.default.on("exit", (worker)=>{
        if (worker.exitedAfterDisconnect) {
            _cluster.default.fork();
        } else {
            process.exit(1);
        }
    });
} else {

// ...
```

If I add the env' variable "__NEXT_DISABLE_MEMORY_WATCHER=1" the problem disappear. But this is not ideal though, memory leaks are not great 😄

**Reply by icyJoseph** [reply]
Right, that looks very hacky to be honest 😅 - prevent memory leaks by killing the CPU. Anyhow, that was good debugging on your side.

I wonder if you could adjust this through the experimental flags.

I also wonder why is it that forking is being problematic here, is it an actual node bug under the hood?

**Reply by Eliepse** [reply]
Ha ha, yes it's... a way to solve a problem 😆 

It does not seems that other Next users had the same issue as me. It's probably related to my config somewhere. The thing is I absolutely don't know what to tweek. I will try a clean install of Nodejs but I'm not sure it will fix the issue.

**Reply by Eliepse** [reply]
So I tried a clean reinstall of Node 18.13 and no luck... Still breaking. 😢

**Comment by Eliepse** [comment, marked as answer]
For anyone that need an answer, the memory leak was caused by _redux-state-sync._ If you have the same issue, check if you also have a dependency who have some memory leak issues ! 😉 

I removed it and now it works like a charm !
